devito/types/tensor.py

```diff
@@ -0,0 +1,598 @@
+                return func(name=name, grid=self.grid, space_order=self.space_order,
+                            components=comps, time_order=to, symmetric=is_symm,
+                            diagonal=is_diag)
+            # All cases should be covered but defaults to sympy if in another case
+            else:
+                return super(TensorFunction, self).__mul__(other)
+        except AttributeError:
+            # If custom product fails defaults to sympy if in another case
+            return super(TensorFunction, self).__mul__(other)
+
+    def __rmul__(self, other):
+        """
+        Right multiplication of a TensorFunction T as T*u where u can be a TensorFunction,
+        a VectorFunction or a Function/scalar.
+        Computes it via mul with tranposes.
+        """
+        other = sympy.sympify(other)
+        try:
+            if other.is_Function:
+                return self._eval_scalar_mul(other)
+            elif other.is_VectorValued or other.is_TensorValued:
+                return (self.T.__mul__(other.T)).T
+            else:
+                return super(TensorFunction, self).__rmul__(other)
+        except AttributeError:
+            return super(TensorFunction, self).__mul__(other)
+
+    def eval_at(self, var):
+        """
+        Evaluate tensor at var location
+        """
+        def entry(i, j):
+            return self[i, j].eval_at(var[i, j])
+        comps = [[entry(i, j) for i in range(self.cols)] for j in range(self.rows)]
+        to = getattr(self, 'time_order', 0)
+        func = tens_func(self, self)
+        return func(name='%s%s' % (self.name, var.name),
+                    grid=self.grid, space_order=self.space_order,
+                    components=comps, time_order=to, symmetric=self.is_symmetric,
+                    diagonal=self.is_diagonal)
+
+    @classmethod
+    def __dtype_setup__(cls, **kwargs):
+        return Function.__dtype_setup__(**kwargs)
+
+    @classmethod
+    def __indices_setup__(cls, **kwargs):
+        return Function.__indices_setup__(grid=kwargs.get('grid'),
+                                          dimensions=kwargs.get('dimensions'))
+
+    @property
+    def is_diagonal(self):
+        return self._is_diagonal
+
+    @property
+    def is_symmetric(self):
+        return self._is_symmetric
+
+    @property
+    def indices(self):
+        return self._indices
+
+    @property
+    def staggered(self):
+        return self._staggered
+
+    @property
+    def space_dimensions(self):
+        return self.indices
+
+    @property
+    def grid(self):
+        return self._grid
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def space_order(self):
+        return self._space_order
+
+    @property
+    def evaluate(self):
+        return self.applyfunc(lambda x: x.evaluate)
+
+    # Custom repr and str
+    def __str__(self):
+        name = "SymmetricTensor" if self._is_symmetric else "Tensor"
```

Comment:
Wouldn't special classes be handy for these cases ? we could return them through `__new__`...